packages/maskbook/src/plugins/ITO/hooks/usePoolPayload.ts

```diff
@@ -1,26 +1,34 @@
 import { PluginITO_RPC } from '../messages'
 import type { JSON_PayloadInMask } from '../types'
+import { useUpdate } from 'react-use'
+import { unstable_useTransition } from 'react'
 
-const storage = new Map<string, JSON_PayloadInMask>()
-
+const cache = new Map<string, [0, Promise<void>] | [1, JSON_PayloadInMask] | [2, Error]>()
+export function poolPayloadErrorRetry() {
+    cache.forEach(([status], pid) => status === 2 && cache.delete(pid))
+}
 export function usePoolPayload(pid: string) {
-    if (!storage.has(pid)) throw suspender(pid)
-    return {
-        payload: storage.get(pid)!,
-        retry: () => retry(pid),
+    const rec = cache.get(pid)
+    const forceUpdate = useUpdate()
+    const [startTransition] = unstable_useTransition({ busyDelayMs: 1000 })
+    if (!rec) {
+        const p = suspender(pid)
+            .then((val) => void cache.set(pid, [1, val]))
+            .catch((e) => void cache.set(pid, [2, e]))
+        cache.set(pid, [0, p])
+        throw p
     }
+    if (rec[0] === 1)
+        return {
+            payload: rec[1],
+            retry: () =>
+                startTransition(() => {
```

Comment:
If you think we should show the loading state directly you can remove the `startTransition`. BTW `useTransition` returns a `isPending` argument so you can return it as `isRefeshing` and still show the old data with a loading refresh hint

Reply:
Hmm wait, I think this code is not ready for `startTransition` cause it is using the global state (rely on `cache`) so remove it if you like. I'll provide an infra later that fully works with these whole bunch of things